Signal/src/ViewControllers/DebugUI/DebugUIDiskUsage.m

```diff
@@ -0,0 +1,299 @@
+            if (error) {
+                OWSFail(@"Could not remove orphan file at: %@", filePath);
+            }
+        }
+    }
+}
+
++ (void)printPaths:(NSArray<NSString *> *)paths label:(NSString *)label
+{
+    for (NSString *path in [paths sortedArrayUsingSelector:@selector(compare:)]) {
+        DDLogError(@"%@: %@", label, path);
+    }
+}
+
++ (void)saveAllAttachments
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    [storageManager.newDatabaseConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+
+        NSMutableArray<TSAttachmentStream *> *attachmentStreams = [NSMutableArray new];
+        [transaction enumerateKeysAndObjectsInCollection:TSAttachmentStream.collection
+                                              usingBlock:^(NSString *key, TSAttachment *attachment, BOOL *stop) {
+                                                  if (![attachment isKindOfClass:[TSAttachmentStream class]]) {
+                                                      return;
+                                                  }
+                                                  TSAttachmentStream *attachmentStream
+                                                      = (TSAttachmentStream *)attachment;
+                                                  [attachmentStreams addObject:attachmentStream];
+                                              }];
+
+        DDLogInfo(@"Saving %zd attachment streams.", attachmentStreams.count);
+
+        // Persist the new localRelativeFilePath property of TSAttachmentStream.
+        // For performance, we want to upgrade all existing attachment streams in
+        // a single transaction.
+        for (TSAttachmentStream *attachmentStream in attachmentStreams) {
+            [attachmentStream saveWithTransaction:transaction];
+        }
+    }];
+}
+
++ (void)deleteOldMessages_3Months
+{
+    NSTimeInterval kMinute = 60.f;
+    NSTimeInterval kHour = 60 * kMinute;
+    NSTimeInterval kDay = 24 * kHour;
+    NSTimeInterval kMonth = 30 * kDay;
+    [self deleteOldMessages:kMonth * 3];
+}
+
++ (void)deleteOldMessages:(NSTimeInterval)maxAgeSeconds
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    [storageManager.newDatabaseConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+
+        NSMutableArray<NSString *> *threadIds = [NSMutableArray new];
+        YapDatabaseViewTransaction *interactionsByThread = [transaction ext:TSMessageDatabaseViewExtensionName];
+        [interactionsByThread enumerateGroupsUsingBlock:^(NSString *group, BOOL *stop) {
+            [threadIds addObject:group];
+        }];
+        NSMutableArray<TSInteraction *> *interactionsToDelete = [NSMutableArray new];
+        for (NSString *threadId in threadIds) {
+            [interactionsByThread enumerateKeysAndObjectsInGroup:threadId
```

Comment:
The advantage to doing this per-thread is that we can `stop` when we find a "new" message, since this view is sorted.  

This functionality too is something that'd be nice to do in a job started on launch, if/when we add a "cull messages after N days" setting.  If/when we do that in prod, this kind of perf improvement will be key.